homeassistant/components/fan/__init__.py

```diff
@@ -196,10 +504,81 @@ def state_attributes(self) -> dict:
 
         if supported_features & SUPPORT_SET_SPEED:
             data[ATTR_SPEED] = self.speed
+            data[ATTR_PERCENTAGE] = self.percentage
+            data[ATTR_PRESET_MODE] = self.preset_mode
```

Comment:
Can a fan have presets but not speed controls? I vaguely recall there was at least one fan like that 😤 in which case should this be separated out under its own supported feature?

Reply:
I could add a new flag and have it only take effect once they have implemented `set_percentage` or `set_preset_mode` in their integration, otherwise it will look at `SUPPORT_SET_SPEED` for backwards compatibility.

Reply:
That's going to be an odd behavior change though when they do the conversion.

For maximum compatibility, we could check if there are `preset_modes` and only show the attribute if there are.

Reply:
I started going down that rabbit hole and I think this would need to come later in another update to the model as it would break backwards compatibility since the integrations would have to add the new `SUPPORT_SET_PRESET_MODE`.

Reply:
We can hide the attribute if there are no entries in `self.preset_modes`. I don't think it will break anything functional (just some tests to update).

Reply:
I was more thinking about the other direction:

https://github.com/home-assistant/core/blob/dev/homeassistant/components/xiaomi_miio/fan.py#L797
https://github.com/home-assistant/core/blob/dev/homeassistant/components/xiaomi_miio/fan.py#L369

A Xiaomi Miio Air Purifier Pro doesnt actually have *any* speeds, *everything* will be in `self.preset_modes`. What happens for one of those? We still have to support "Silent" as a speed for compatibility, I guess. But are there any other gotchas.

Reply:
I'll add a demo fan that has only presets and no speeds to make sure everything is handled in a sane way

Reply:
To support this I'm going to have to add `SUPPORT_PRESET_MODES`.

Reply:
I've added this and a demo fan that only does preset modes.

I may need to tweak a few tests once the ci runs

Reply:
<img width="408" alt="Screen Shot 2021-01-25 at 3 09 33 PM" src="https://user-images.githubusercontent.com/663432/105766479-69e25e80-5f1f-11eb-9855-a0ce7f753749.png">

Reply:
I updated the developer docs pr for this as well

Reply:
Tested on one of my instances and looks good